plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/IGenerator2.java

```diff
@@ -32,24 +33,19 @@
 	 * are non-blocking if the given fsa is implementing that pattern.
 	 * Read operations will happen in the background but
 	 * are blocking.
-	 * 
-	 * @param input
-	 *            the input for which to generate artefacts
-	 * @param fsa
-	 *            file system access to be used to generate files
 	 */
-	void doGenerate(Resource input, IFileSystemAccess2 fsa);
+	void doGenerate(Resource input, IFileSystemAccess2 fsa, CancelIndicator cancelIndicator);
 
 	/**
 	 * Before the generation is triggered, the resource or the entire resource set
-	 * may be prepared such that no modification will happen while {@link #doGenerate(Resource, IFileSystemAccess2)}
+	 * may be prepared such that no modification will happen while {@link #doGenerate(Resource, IFileSystemAccess2, CancelIndicator)}
 	 * is executed.
 	 */
-	void beforeGenerate(Resource input, IFileSystemAccess2 fsa);
+	void beforeGenerate(Resource input, IFileSystemAccess2 fsa, CancelIndicator cancelIndicator);
 
 	/**
 	 * Release any data that is no longer necessary after the generator ran. 
 	 */
-	void afterGenerate(Resource input, IFileSystemAccess2 fsa);
+	void afterGenerate(Resource input, IFileSystemAccess2 fsa, CancelIndicator cancelIndicator);
```

Comment:
We should document the contract of the life cycle hooks in case of cancellation. E.g. afterGenerate is always called independently from the cancellation state of the generator or smth like that.